src/timely-util/src/builder_async.rs

```diff
@@ -0,0 +1,344 @@
+            shared_frontiers: Rc::clone(&self.shared_frontiers),
+            reactor_registry: Rc::downgrade(&self.registered_wakers),
+            index,
+        }
+    }
+
+    /// Adds a new output, returning the output handle and stream.
+    pub fn new_output<D: Container>(
+        &mut self,
+    ) -> (
+        OutputWrapper<G::Timestamp, D, TeeCore<G::Timestamp, D>>,
+        StreamCore<G, D>,
+    ) {
+        self.builder.new_output()
+    }
+
+    /// Adds a new output with connetion information, returning the output handle and stream.
+    ///
+    /// The `connection` parameter contains promises made by the operator for each of the existing
+    /// *inputs*, that any timestamp appearing at the input, any output timestamp will be greater
+    /// than or equal to the input timestamp subjected to a `Summary` greater or equal to some
+    /// element of the corresponding antichain in `connection`.
+    ///
+    /// Commonly the connections are either the unit summary, indicating the same timestamp might
+    /// be produced as output, or an empty antichain indicating that there is no connection from
+    /// the input to the output.
+    pub fn new_output_connection<D: Container>(
+        &mut self,
+        connection: Vec<Antichain<<G::Timestamp as Timestamp>::Summary>>,
+    ) -> (
+        OutputWrapper<G::Timestamp, D, TeeCore<G::Timestamp, D>>,
+        StreamCore<G, D>,
+    ) {
+        self.builder.new_output_connection(connection)
+    }
+
+    /// Creates an operator implementation from supplied logic constructor.
+    pub fn build<B, L>(self, constructor: B)
+    where
+        B: FnOnce(Vec<Capability<G::Timestamp>>) -> L,
+        L: Future + 'static,
+    {
+        let operator_waker = self.operator_waker;
+        let registered_wakers = self.registered_wakers;
+        let shared_frontiers = self.shared_frontiers;
+        self.builder.build_reschedule(move |caps| {
+            let mut logic_fut = Some(Box::pin(constructor(caps)));
+            move |new_frontiers| {
+                // First, discover if there are any frontier notifications
+                {
+                    let mut old_frontiers = shared_frontiers.borrow_mut();
+                    for ((old, pending), new) in old_frontiers.iter_mut().zip(new_frontiers) {
+                        if PartialOrder::less_than(&old.borrow(), &new.frontier()) {
+                            *old = new.frontier().to_owned();
+                            *pending = true;
+                        }
+                    }
+                }
+                // Then, wake up any registered Wakers for the input streams while taking care to
+                // not reactivate the timely operator since that would lead to an infinite loop.
+                // This ensures that the input handles wake up properly when managed by other
+                // executors, e.g a `select!` that provides its own Waker implementation.
```

Comment:
`select` doesn't provide its own, I dont think, but `FuturesUnordered` does? We can merge without this, but this logic is complex, so we should add a test as a followup

Reply:
You're right! I thought it did so that it wakes up the right branch. `FuturesUnordered` does though, I'll change the comment to refer to that instead

Reply:
WOW `FuturesUnordered`'s impl is so complex...but yeah I think it does use custom wakers